### PR TITLE
✨💥 Add login functionality with either username or email

### LINKED ISSUE #587: Add username-password login option
**Is your feature request related to a problem? Please describe.**
It would be great if träwelling would allow users to sign in with username-passwort alongside email-password like Twitter and other networks.

**Additional context**
For consistency this feature should be available for the API as well.

### PR DESCRIPTION
**Pull request by HerrLevin**
Fixes #587

### CODE REVIEW

File: app/Http/Controllers/API/v1/AuthController.php [hunk 1]
```diff
@@ -52,34 +51,27 @@ public function register(Request $request): JsonResponse {
...
+     *
      * @return JsonResponse
      * @api v1
      */
     public function login(Request $request): JsonResponse {
-        $validator = Validator::make($request->all(), [
-            'email'    => ['required', 'email'],
-            'password' => ['required']
-        ]);
+        $validated = $request->validate(['login' => ['required'], 'password' => ['required']]);
```
Comment (MrKrisKrisu) [inline app/Http/Controllers/API/v1/AuthController.php]: Maybe we can add a `max:255` to login. E-Mails are limited to 255 and usernames to 25.

Comment (MrKrisKrisu) [reply app/Http/Controllers/API/v1/AuthController.php]: and `min:8` for password would be nice, too.

File: app/Http/Controllers/API/v1/AuthController.php [hunk 2]
```diff
@@ -52,34 +51,27 @@ public function register(Request $request): JsonResponse {
...
-                                        'token'      => $token->accessToken,
-                                        'expires_at' => $token->token->expires_at->toIso8601String()
-                                    ], 200)
-                             ->header('Authorization', $token->accessToken);
+            return $this->sendv1Response(['token'      => $token->accessToken,
+                                          'expires_at' => $token->token->expires_at->toIso8601String()])
+                        ->header('Authorization', $token->accessToken);
         }
-        return response()->json(['error' => 'login_error'], 401);
+        return $this->sendError('Not matching credentials', 401);
```
Comment (MrKrisKrisu) [inline app/Http/Controllers/API/v1/AuthController.php]:
```suggestion
        return $this->sendError('Non-matching credentials', 401);
```

File: app/Http/Controllers/Auth/LoginController.php [hunk 3]
```diff
@@ -38,6 +40,22 @@ public function __construct() {
         $this->middleware('guest')->except('logout');
     }
 
+    public function login(Request $request): RedirectResponse {
+        $validated = $request->validate(['login'    => 'required',
+                                         'password' => 'required',
```
Comment (MrKrisKrisu) [inline app/Http/Controllers/Auth/LoginController.php]: Maybe we can add a `max:255` to login. E-Mails are limited to 255 and usernames to 25.
and `min:8` for password would be nice, too.

File: app/Http/Controllers/Auth/LoginController.php [hunk 4]
```diff
@@ -38,6 +40,22 @@ public function __construct() {
...
+                                         'remember' => 'nullable|in:1']);
```
Comment (MrKrisKrisu) [inline app/Http/Controllers/Auth/LoginController.php]:
```suggestion
                                         'remember' => ['nullable', 'in:1']]);
```

File: app/Http/Controllers/Backend/Auth/LoginController.php [hunk 5]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+namespace App\Http\Controllers\Backend\Auth;
+
+use App\Http\Controllers\Controller;
+use Illuminate\Support\Facades\Auth;
+
+class LoginController extends Controller
+{
+    public static function login($login, $password, $remember = false): bool {
```
Comment (MrKrisKrisu) [inline app/Http/Controllers/Backend/Auth/LoginController.php]: Type of parameters?

File: app/Http/Controllers/Backend/Auth/LoginController.php [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+        $login_type = filter_var($login, FILTER_VALIDATE_EMAIL) ? 'email' : 'username';
+
+        if (Auth::attempt([$login_type => $login, 'password' => $password], $remember)) {
```
Comment (MrKrisKrisu) [inline app/Http/Controllers/Backend/Auth/LoginController.php]:
```suggestion
        $loginType = filter_var($login, FILTER_VALIDATE_EMAIL) ? 'email' : 'username';

        if (Auth::attempt([$loginType => $login, 'password' => $password], $remember)) {
```

**Review by MrKrisKrisu — CHANGES_REQUESTED** [review]

**Comment by MrKrisKrisu** [comment]
We should then wait with the release until @Vogeslu and @jheubuch have adapted their apps accordingly.

**Comment by jheubuch** [comment]
Fine from my side 👍🏼

File: app/Http/Controllers/API/v1/AuthController.php (hunk 1 above)
Comment (HerrLevin) [reply app/Http/Controllers/API/v1/AuthController.php]: Idk, if that's really necessary since it'll return false either way. 😅

**Review by MrKrisKrisu — APPROVED** [review]

**Pull request merged**